cmd/action/internal/update/update.go

```diff
@@ -0,0 +1,150 @@
+// 2. If last poll was more than 24h, poll GitHub public API https://docs.github.com/en/rest/reference/releases#get-the-latest-release
+// 3. Store the latest release metadata
+// 4. If current build Version, that is not development, is lower than the latest release, notify user
+func CheckForUpdate(version string, logF func(i ...interface{})) {
+	if !enabled(version) {
+		return
+	}
+	p, err := homedir.Expand("~/.config/atlas")
+	if err != nil {
+		return
+	}
+	ok, message, err := shouldUpdate(version, p, latestReleaseFromGithub)
+	if err != nil || !ok {
+		return
+	}
+	logF(message)
+}
+
+func enabled(version string) bool {
+	if _, ok := os.LookupEnv("ATLAS_NO_UPDATE_NOTIFIER"); ok {
+		return false
+	}
+	if _, ok := os.LookupEnv("GITHUB_ACTIONS"); ok {
+		return false
+	}
+	if ok := semver.IsValid(version); !ok {
+		return false
+	}
+	return true
+}
+
+func shouldUpdate(version string, path string, latestReleaseF func() (LatestRelease, error)) (bool, string, error) {
+	r, _ := localStore(path)
+	if shouldSkip(r) {
+		return false, "", nil
+	}
+	l, err := latestReleaseF()
+	if err != nil {
+		return false, "", err
+	}
+	if err := saveStore(path, l, time.Now()); err != nil {
+		return false, "", err
+	}
+	if semver.Compare(version, l.Version) != -1 {
+		return false, "", nil
+	}
+	return true, fmt.Sprintf("new atlas version %s, %s", l.Version, l.URL), nil
+}
+
+func latestReleaseFromGithub() (LatestRelease, error) {
+	req, err := http.NewRequest("GET", "https://api.github.com/repos/ariga/atlas/releases/latest", nil)
+	if err != nil {
+		return LatestRelease{}, err
+	}
+	req.Header.Set("Content-Type", "application/json; charset=utf-8")
+	req.Header.Set("Accept", "application/vnd.github.v3+json")
+	// https://docs.github.com/en/rest/overview/resources-in-the-rest-api#user-agent-required
+	req.Header.Set("User-Agent", "Ariga-Atlas-CLI")
+	client := &http.Client{Timeout: time.Second * 10}
+	resp, err := client.Do(req)
+	if err != nil {
+		return LatestRelease{}, err
+	}
+	defer resp.Body.Close()
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return LatestRelease{}, err
+	}
+	ok := resp.StatusCode >= 200 && resp.StatusCode < 300
+	if !ok {
+		return LatestRelease{}, fmt.Errorf("update: failed to fetch latest release version")
+	}
+	var b LatestRelease
+	err = json.Unmarshal(data, &b)
+	if err != nil {
+		return LatestRelease{}, err
+	}
+	return b, nil
+}
+
+func shouldSkip(r *Store) bool {
+	if r == nil {
+		return false
+	}
+	return time.Since(r.CheckedAt).Hours() < 24
+}
+
+func localStore(path string) (*Store, error) {
+	b, err := ioutil.ReadFile(fileLocation(path))
+	if err != nil {
+		return nil, err
+	}
+	var s Store
+	err = toml.Unmarshal(b, &s)
+	if err != nil {
+		return nil, err
+	}
+	return &s, nil
+}
+
+func saveStore(path string, l LatestRelease, t time.Time) error {
+	s := Store{Version: l.Version, URL: l.URL, CheckedAt: t}
+	err := os.MkdirAll(path, 0755)
```

Comment:
Why not `os.MkdirAll(dir, os.ModePerm)`. here you limit the write to the owner.

Reply:
the cli is stored in user permission in the user's home

Reply:
checked other CLI's in my home folder all are 0755